src/mds/ScrubStack.cc

```diff
@@ -436,3 +498,140 @@ void ScrubStack::_validate_inode_done(CInode *in, int r,
 
 ScrubStack::C_KickOffScrubs::C_KickOffScrubs(MDCache *mdcache, ScrubStack *s)
   : MDSInternalContext(mdcache->mds), stack(s) { }
+
+void ScrubStack::complete_control_contexts(int r) {
+  ceph_assert(mdcache->mds->mds_lock.is_locked_by_me());
+
+  for (auto ctx : control_ctxs) {
+    ctx->complete(r);
+  }
+  control_ctxs.clear();
+}
+
+void ScrubStack::set_state(State next_state) {
+    if (state != next_state) {
+      dout(20) << __func__ << ", from state=" << state << ", to state="
+               << next_state << dendl;
+      state = next_state;
+    }
+}
+
+bool ScrubStack::scrub_in_transition_state() {
+  ceph_assert(mdcache->mds->mds_lock.is_locked_by_me());
+  dout(20) << __func__ << ": state=" << state << dendl;
+
+  // STATE_RUNNING is considered as a transition state so as to
+  // "delay" the scrub control operation.
+  if (state == STATE_RUNNING || state == STATE_PAUSING) {
+    return true;
+  }
+
+  return false;
+}
+
+void ScrubStack::scrub_status(std::ostream *ss) {
+  ceph_assert(mdcache->mds->mds_lock.is_locked_by_me());
+  ceph_assert(ss != nullptr);
+
+  if (state == STATE_NOT_RUNNING) {
+    *ss << "no active scrubs running";
+  } else if (state == STATE_RUNNING) {
+      *ss << "scrub active (" << stack_size << " inodes in the stack)";
```

Comment:
I think clear_inode_stack can be true in running state.

Reply:
you are correct -- I'll rearrange the status part.

regarding having clear_inode_stack as a bit in a state (STATE_RUNNING), I think, if there are other possible operations that have to be performed in STATE_RUNNING, keeping a bitmap of operations would be useful. Moreover, it would be more useful if we can keep a `state -> operation bitmap` mapping to make it more generic. Right now, we just have abort operation, so using a bool does the job.